proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,235 @@
+
+In the new stake rewards distribution approach, we will separate the computation of
+rewards from the actual distribution of rewards at the epoch boundary by
+dividing the process into two distinct phases: (1) rewards calculation, and (2)
+rewards distribution.
+
+The "reward calculation" phase is defined as the blocks that fall under it are
+those where time is allocated to calculating the rewards, while the "rewards
+distribution" phase is defined as the blocks that fall under it are those where
+time is allocated to distributing the rewards.
+
+### Rewards Calculation
+
+The reward computation phase lasts for first `N` block heights since the
+beginning of the epoch. When reaching block height `N` after the start block
+of the `reward calculation phase`, the bank will mark it the end the
+`reward calculation phase`. A conservative value of 1,000 is chosen for `N` to
+accommodate the existing validators on Solana Mainnet Beta.
+
+At this point, the reward computation result should be available. In case that
+the reward computation result is not available yet at the `N_th_` block-height,
+the bank runtime will block and wait until the reward computation is finished, before
+processing any blocks with height greater than `N`.
+
+At the end of the reward computation phase (end of block height `N`),
+`EpochRewardHistory` sysvar account and a set of `RewardReserve` sysvar
+accounts, which helps to track and verify the reward distribution progress, will be
+updated.
+
+### `EpochRewardHistory` Sysvar Account
+
+`EpochRewardHistory` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardHistory` is an array of `RewardHistoryEntry`,
+where each entry corresponds to the reward for one particular epoch. The
+maximum depth of the history are chosen to the same as the depth of stake history for consistency, which
+is 512.
+
+The layout of `EpochRewardHistory` sysvar is shown in the following pseudo code.
+
+```
+struct RewardHistoryEntry {
+   total_reward_in_lamport: u64,          // total rewards for this epoch
+   distributed_reward_in_lamport: u64,    // already distributed reward amount
+   number_reward_reserve_accounts: u32,   // number of reward reserve accounts for reward distribution
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+The `EpochRewardHistory` is updated at the first block of the epoch boundary,
+as both the total epoch rewards and vote account rewards become available at
+this time. Once the vote account rewards have been distributed, a
+`RewardHistoryEntry` will be added to the `EpochRewardHistory` array with the
+total amount of reward, a distributed value equal to the sum of the vote
+account rewards, and the number of reward reserve accounts for stake reward
+distribution. If the length of the history array reaches the maximum depth (512), the
+oldest entry will be removed from the array (i.e., "popped").
```

Comment:
During this first block of the epoch, the M `EpochRewardReserve` accounts would also need to be created as well, if that's where the new reward lamports are going to be held until distribution (to keep the bank capitalization checks passing).

We should know M at this time because the total number of stake accounts is known and we distribute at a fixed rate of 64 stake accounts per tick